src/timezone_alternative_b.rs

```diff
@@ -0,0 +1,180 @@
+#![allow(dead_code, unreachable_pub)]
+
+use crate::offset::FixedOffset;
+use crate::Days;
+use crate::Months;
+use crate::NaiveDateTime;
+use crate::ParseResult;
+use crate::TimeDelta;
+use crate::Utc;
+
+// this is nice because it avoids the type paramter.
+// alternatively the `TimeZoneManager` could have an assocaited `TimeZone` ZST or equivalent that this is parametized by
+#[derive(Clone, Copy)]
+pub struct DateTime {
+    datetime: NaiveDateTime,
+    offset: FixedOffset,
+    // could potentially include some information on the timezone name here to allow `%Z` style formatting
+}
```

Comment:
one option would be having a function on the `TimeZoneManager` that 'normalises' or 'localises' the offset, but the main way the API is intended to be used is something like, `my_timezone_manager.add(my_datetime, Duration::from_secs(24 * 60 * 60))`

Potentially another option that could be more ergonomic is:
`my_datetime.add(&my_timezone_manager, Duration::from_secs(24 * 60 * 60))`

Reply:
I'm not sure I'm convinced by either approach; the Datetime is essentially losing its TimeZone implementation and instance, and application code needs to keep both around for any kind of computation.

I think alternative a is a lot cleaner in that sense.

Reply:
Fair call. We could do the following in the `TimeZoneManager` style:

have the offset field be a type parameter, with 

```rust 
pub struct DateTime<Offset=FixedOffset> {
    datetime: NaiveDateTime,
    offset: Offset,
}
```

then when using a given `TimeZoneManager` it would operate on `DateTime`s with the relevant `Offset` type. This ties them together and either discourages or prevents adding time without the associated timezone information to update the offset. (Even in the prevent option, users could simply extract the `NaiveDateTime` or change the `Offset` if desired)

The extra benefit of making `Offset` a type parameter is that in the case of a named timezone, it can store the name internally, similar to `chrono_tz`'s offset type.